flocker/node/agents/cinder.py

```diff
@@ -67,30 +71,42 @@ def _run_with_logging(self, *args, **kwargs):
 
         # See https://clusterhq.atlassian.net/browse/FLOC-2054
         # for ensuring all method arguments are serializable.
-        with OPENSTACK_ACTION(operation=[method_name, args, kwargs]):
-            try:
-                return method(*args, **kwargs)
-            except NovaClientException as e:
-                NOVA_CLIENT_EXCEPTION(
-                    code=e.code,
-                    message=e.message,
-                    details=e.details,
-                    request_id=e.request_id,
-                    url=e.url,
-                    method=e.method,
-                ).write()
-                raise
-            except KeystoneHttpError as e:
-                KEYSTONE_HTTP_ERROR(
-                    code=e.http_status,
-                    message=e.message,
-                    details=e.details,
-                    request_id=e.request_id,
-                    url=e.url,
-                    method=e.method,
-                    response=e.response.text,
-                ).write()
-                raise
+        retry = True
+        attempt = 0
+        while (retry and attempt < MAX_ATTEMPTS):
+            with OPENSTACK_ACTION(operation=[attempt, method_name, args, kwargs]):
+                try:
+                    return method(*args, **kwargs)
+                except KeystoneOverLimit as e:
+                    retry_after = int(e.retry_after)
+                    time.sleep(retry_after)
+                    retry = True
+                    attempt += 1
+                    if attempt == MAX_ATTEMPTS:
+                        raise
```

Comment:
I'm fairly tempted to suggest this retry logic belongs in a separate decorator, not in the logging decorator.  Maybe that refactoring isn't the highest priority task now - on the other hand, the retry logic should really have unit tests.  This code currently doesn't have direct tests for the retry behavior and it will be a lot easier to write them if you can exercise the retry code without all of this OpenStack baggage attached to it.

So .. yea.  I think it would be better to leave the logging decorator alone and create a new `interface_decorator` for this automatic sleep-and-retry behavior.  Parameterize enough of the pieces (apart from the method, I'm thinking about the exception type which prompts a retry) so that you can write some generic unit tests that don't interact with OpenStack at all.  Then you can apply that decorator in the same place `auto_openstack_logging` is being applied.

I'd also suggest adding `KeystoneOverLimit` to the logging decorator and applying the retry decorator "outside of" ("on top of"? eh) the logging decorator so that each over-limit exception that might prompt a retry also gets logged.